posthog/helpers/item_history.py

```diff
@@ -0,0 +1,150 @@
+import dataclasses
+from typing import Dict, Iterable, List, Optional, Tuple, Union
+
+from posthog.models import HistoricalVersion
+from posthog.models.utils import UUIDT
+
+
+@dataclasses.dataclass(frozen=True)
+class HistoryListItem:
+    email: Optional[str]
+    name: Optional[str]
+    user_id: int
+    action: str
+    detail: Dict[str, Union[int, str, Dict]]
+    created_at: str
+
+
+def compute_history(
+    history_type: str,
+    item_id: int,
+    version_pairs: Iterable[Tuple[HistoricalVersion, Optional[HistoricalVersion]]],
+    team_id: Optional[int] = None,
+    organization_id: Optional[UUIDT] = None,
+):
+    """
+    TODO Purposefully leaving this as unstructured "Arrow code" to get to "shameless green"
+    TODO until there are at least two or three types having history computed
+    TODO to avoid premature abstraction
+
+    takes a type of item e.g. FeatureFlag or Insight
+    and a set of version pairs
+    and uses them to compute a history for that instance of that type
+
+    The version pairs are an overlapping zip of the item's history
+    So if its history is [a, b, c, d] this function receives [(a, b), (b, c), (c, d), (d, None)]
+    (where a is the most recent recorded change)
+    It uses the right-hand side of each tuple to determine what changed to generate the left-hand side
+    """
+    history: List[HistoryListItem] = []
+
+    _process_pairs(history, history_type, version_pairs)
+
+    if len(history) == 0:
+        imported_version = HistoricalVersion(
+            state={"key": "unknown"},
+            name="FeatureFlag",
+            action="update",
+            item_id=item_id,
+            created_by_name="history hog",
+            created_by_email="history.hog@posthog.com",
+            created_by_id=0,
+        )
+        if team_id:
+            imported_version.team_id = team_id
+        elif organization_id:
+            imported_version.organization_id = organization_id
+        imported_version.save()
```

Comment:
This code looks new. Is it necessary? I tried locally, and I'm not sure what this did, as all histories for all flags that I tried were empty. I'm also not sure we want to add something to the database in a function that's supposed to just compute something. 

Is there a backend reason for an event like this? If not, I'd be in favour of the "no surprises" approach and just remove this. We can improve the "empty list" view in the frontend perhaps?

Reply:
>  I'm also not sure we want to add something to the database in a function that's supposed to just compute something.

Good shout. Was focussed on behaviour not structure. But there shouldn't be a side effect in there!

Reply:
The reason to import the history is to capture state of items that exist before we release history logging. That way when someone updates something they get

![Screenshot 2022-02-16 at 11 43 04](https://user-images.githubusercontent.com/984817/154257924-b440a0fd-d358-4079-8a09-71191099d29b.png)

Otherwise we wouldn't be able to see what changes had been made in this update and every field shows as an update.

Reply:
...to capture state of items *without a migration being needed*